### PR TITLE
Back to union storage

### LINKED ISSUE #30: Dealing with extra parameters in recursive sumtypes
Hey @MasonProtter, really nice package.

Is there a way I can make a recursive SumType which uses the correct parametrized field type?

For example, say I want to make a linked list, with a separate type for the end of the list:

```julia
@sum_type LinkedList begin
    BranchNode(val::Float64, child::LinkedList)
    EndNode(val::Float64)
end
```

However, this `LinkedList` type referenced in `BranchNode` is actually generic and will incur type inference issues, as the sum type is parametrized:

```julia
julia> typeof(BranchNode(0.1, EndNode(0.1)))
LinkedList{8, 1, UInt8}
```

Is there a way I can get this same parametrization within the `@sum_type` macro, so that I may use it to build recursive SumTypes?

Cheers,
Miles

### PR DESCRIPTION
**Pull request by MasonProtter**
This undoes the major overhaul of version 0.4 where we were doing computed storage. With this change, a bunch of things get a lot simpler, and we don't need to carry around extra parameters in the sum type.

While the computed storage did benefit some situations, julia's `Union` is also quite good. For all `isbits` data it should have the same performance profile as before, but for non-`isbits` data, there are some tradeoffs as to what ends up heap allocated and what doesn't, and I think that `Union` actually does a better job when you're modifying and interacting with the data since there's usually fewer pointers involved.

Fixes #30

### CODE REVIEW

**Comment by MasonProtter** [comment]
Fixes #30

**Comment by MasonProtter** [comment]
@Krastanov, I know you have some code relying on SumTypes.jl for performance referenced in https://github.com/MasonProtter/SumTypes.jl/issues/28. Could I ask you to test out this PR and see if it causes any performance regressions (or improvements!) for you?

File: README.md [hunk 1]
```diff
@@ -334,58 +295,45 @@ julia> SumTypes.show_sumtype(io::IO, ::MIME"text/plain", x::Fruit2) = @cases x b
 julia> apple
 apple!
 ```
-If you overload `Base.show` directly inside a package, you might get annoying method deletion warnings during pre-compilation.
+If you overload `Base.show** directly inside a package, you might get annoying method deletion warnings during pre-compilation.
```
Comment (mfiano) [inline README.md:298]: Ugh meant for that to be a comment, not a review. Excuse me while I finish waking up.

**Comment by Krastanov** [comment]
The only interesting benchmarks I have are with `isbits` objects, so maybe this is not too informative, but here are the results:

- most benchmarks, especially the ones that are NOT microbenchmarks, show no difference
- these two microbenchmarks are much faster with the newer 0.5.0-dev than with 0.4.8-master: https://github.com/QuantumSavory/QuantumClifford.jl/blob/544ba079c55dea93946cadba2a3aa6cd5d72adca/benchmark/benchmarks.jl#L132-L133 - now they take about 2/3 of the time they used to take. Given your commit message, I guess change in performance with `isbits` is not expected, so I reran the benchmarks and the results persisted.

Tested on:

```
julia> versioninfo()
Julia Version 1.9.1
Commit 147bdf428cd (2023-06-07 08:27 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 32 × AMD Ryzen 9 7950X 16-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, znver3)
  Threads: 32 on 32 virtual cores
```

If you want to double check things for yourself, these are the benchmarks that depend on sumtypes:

https://github.com/QuantumSavory/QuantumClifford.jl/blob/544ba079c55dea93946cadba2a3aa6cd5d72adca/benchmark/benchmarks.jl#L131-L138

**Comment by MasonProtter** [comment]
Well that’s quite encouraging news! I guess julia’s handling of enclosed unions here is a bit smarter than what I was doing and enabled some speedups.

**Pull request merged**